modules/promotion/src/Element/CouponRedemptionForm.php

```diff
@@ -0,0 +1,225 @@
+    ];
+  }
+
+  /**
+   * Builds the coupon redemption form.
+   *
+   * @param array $element
+   *   The form element being processed.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   * @param array $complete_form
+   *   The complete form structure.
+   *
+   * @throws \InvalidArgumentException
+   *   Thrown when the #operation or #default_value properties are empty, or
+   *   when the #default_value property is not a valid entity.
+   *
+   * @return array
+   *   The processed form element.
+   */
+  public static function processForm(array $element, FormStateInterface $form_state, array &$complete_form) {
+    if (!$element['#order'] instanceof OrderInterface) {
+      return $element;
+    }
+
+    $id_prefix = implode('-', $element['#parents']);
+    // @todo We cannot use unique IDs, or multiple elements on a page currently.
+    // @see https://www.drupal.org/node/2675688
+    // $wrapper_id = Html::getUniqueId($id_prefix . '-ajax-wrapper');
+    $wrapper_id = $id_prefix . '-ajax-wrapper';
+
+    $form_state->set('order', $element['#order']);
+    $element = [
+      '#tree' => TRUE,
+      '#prefix' => '<div id="' . $wrapper_id . '">',
+      '#suffix' => '</div>',
+      // Pass the id along to other methods.
+      '#wrapper_id' => $wrapper_id,
+    ] + $element;
+    $element['code'] = [
+      '#type' => 'textfield',
+      '#title' => t('Coupon code'),
+      '#description' => t('Enter your coupon code here.'),
+    ];
+    $element['apply'] = [
+      '#type' => 'submit',
+      '#value' => t('Apply coupon'),
+      '#name' => 'apply_coupon',
+      '#limit_validation_errors' => [
+        array_merge($element['#parents'], ['code']),
+      ],
+    ];
+
+    return $element;
+  }
+
+  /**
+   * Validates the coupon redemption form.
+   *
+   * @param array $element
+   *   The form element.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   */
+  public static function validateForm(array &$element, FormStateInterface $form_state) {
+    $coupon_parents = array_merge($element['#parents'], ['code']);
+    $coupon_code = $form_state->getValue($coupon_parents);
+    if (empty($coupon_code)) {
+      return;
+    }
+    $code_path = implode('][', $coupon_parents);
+    /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
+    $order = $form_state->get('order');
+    $entity_type_manager = \Drupal::entityTypeManager();
+
+    // Validate that the coupon code exists.
+    $coupon = static::loadCouponByCode($coupon_code);
+    if (empty($coupon)) {
+      $form_state->setErrorByName($code_path, t('Coupon is invalid'));
+      return;
+    }
+
+    foreach ($order->get('coupons') as $item) {
+      if ($item->target_id == $coupon->id()) {
+        $form_state->setErrorByName($code_path, t('Coupon is already redeemed'));
+        return;
+      }
+    }
+
+    // If it does, ensure its promotion is still valid.
```

Comment:
"If it does, ensure its promotion is still valid." is not above the code that actually does it, so I was confused.